server/service/software_installers.go

```diff
@@ -32,18 +32,155 @@ type uploadSoftwareInstallerRequest struct {
+	InstallScript     *string
+	PreInstallQuery   *string
+	PostInstallScript *string
+	UninstallScript   *string
+	SelfService       *bool
+}
+
 type uploadSoftwareInstallerResponse struct {
 	Err error `json:"error,omitempty"`
 }
 
 // MaxSoftwareInstallerSize is the maximum size allowed for software
-// installers. This is enforced by the endpoint that uploads installers.
+// installers. This is enforced by the endpoints that upload installers.
 const MaxSoftwareInstallerSize = 500 * units.MiB
 
+// TODO: We parse the whole body before running svc.authz.Authorize.
+// An authenticated but unauthorized user could abuse this.
+func (updateSoftwareInstallerRequest) DecodeRequest(ctx context.Context, r *http.Request) (interface{}, error) {
+	decoded := updateSoftwareInstallerRequest{}
+
+	// populate software title ID since we're overriding the decoder that would do it for us
+	titleID, err := uintFromRequest(r, "id")
+	if err != nil {
+		return nil, badRequestErr("intFromRequest", err)
+	}
+	decoded.TitleID = uint(titleID)
+
+	err = r.ParseMultipartForm(512 * units.MiB)
+	if err != nil {
+		var mbe *http.MaxBytesError
+		if errors.As(err, &mbe) {
+			return nil, &fleet.BadRequestError{
+				Message:     "The maximum file size is 500 MB.",
+				InternalErr: err,
+			}
+		}
+		var nerr net.Error
+		if errors.As(err, &nerr) && nerr.Timeout() {
+			return nil, fleet.NewUserMessageError(
+				ctxerr.New(ctx, "Couldn't upload. Please ensure your internet connection speed is sufficient and stable."),
+				http.StatusRequestTimeout,
+			)
+		}
+		return nil, &fleet.BadRequestError{
+			Message:     "failed to parse multipart form: " + err.Error(),
+			InternalErr: err,
+		}
+	}
+
+	// unlike for uploadSoftwareInstallerRequest, every field is optional, including the file upload
+	if r.MultipartForm.File["software"] != nil || len(r.MultipartForm.File["software"]) > 0 {
+		decoded.File = r.MultipartForm.File["software"][0]
+		if decoded.File.Size > MaxSoftwareInstallerSize {
+			// Should never happen here since the request's body is limited to the maximum size.
+			return nil, &fleet.BadRequestError{
+				Message: "The maximum file size is 500 MB.",
+			}
+		}
+	}
+
+	// default is no team
+	val, ok := r.MultipartForm.Value["team_id"]
+	if ok {
+		teamID, err := strconv.ParseUint(val[0], 10, 32)
+		if err != nil {
+			return nil, &fleet.BadRequestError{Message: fmt.Sprintf("Invalid team_id: %s", val[0])}
+		}
+		decoded.TeamID = ptr.Uint(uint(teamID))
+	}
+
+	val, ok = r.MultipartForm.Value["install_script"]
+	if ok && len(val) > 0 {
+		decoded.InstallScript = &val[0]
+	}
+
+	val, ok = r.MultipartForm.Value["pre_install_query"]
+	if ok && len(val) > 0 {
+		decoded.PreInstallQuery = &val[0]
+	}
+
+	val, ok = r.MultipartForm.Value["post_install_script"]
+	if ok && len(val) > 0 {
+		decoded.PostInstallScript = &val[0]
+	}
+
+	val, ok = r.MultipartForm.Value["uninstall_script"]
+	if ok && len(val) > 0 {
+		decoded.UninstallScript = &val[0]
```

Comment:
Hmm, thanks for bringing this up (and for the playground link you DM'd me). What's weird is this actually seems to work when multiple scripts are changed, so I need to figure out what's going on. I think it works because the pointers are to the index in the multipart array, rather than to the val itself, so it's brittle but functional. Going to hack on the playground example to see if I can repro.

Reply:
Went ahead and renamed values that involve pointers so this is a bit less brittle.